tokstr.h

```diff
@@ -2,30 +2,62 @@
 #define __TOKSTR_H
 
 // tokstr -- textual token iterator with some input independence
+// 2022-01-29 [revised during code review, add regions]
 // 2022-01-25 [initially released inside dnsdbq]
 
-/* example:
+/* example using heap-allocated strings:
 
 	tokstr_t ts = tokstr_string("this:is+-test");
 	for (char *t; (t = tokstr_next(ts, "-:+")) != NULL; free(t))
 		printf("\t\"%s\"\n", t);
 	tokstr_last(&ts);
 
- * will output "this", "is", and "test".
+ * will output "this", "is", and "test". so will this:
+
+	tokstr_t ts = tokstr_string("this:is+-test");
+	for (char t[100]; tokstr_next_copy(ts, "-:+", t, sizeof t) > 0;)
+		printf("\t\"%s\"\n", t);
+	tokstr_last(&ts);
+
+ * as will this:
+
+	tokstr_t ts = tokstr_string("this:is+-test");
+	for (;;) {
+		struct tokstr_reg t = tokstr_next_region(ts, "-:+");
+		if (t.base == NULL)
```

Comment:
using t.base feels like it's using a private field.
I guess you'd need to document in struct tokstr_reg that it's valid for a library user to reference it.  Or add an accessor function?

Reply:
struct tokstr_reg is public. it is used as input to tokstr_region() for example. if we wanted it to be private we would move it out of tokstr.h.

Reply:
please document reading .base
it can't be private (and moved out of tokstr.h) because you return that structure.  Returning a pointer is not appropriate for allocation reasons, but could allow that info hiding.

Reply:
i suspect i had better just document what the structure members do and what base == NULL means. in the zero copy output interface, .base and .size are how each token from the input is described.

Reply:
fixed (b873f91..ddd2a0a)